src/hardware/mixer.cpp

```diff
@@ -2735,7 +2735,12 @@ static bool init_sdl_sound(Section_prop* section)
 		LOG_WARNING("MIXER: SDL changed the requested sample rate of %d to %d Hz",
 		            mixer.sample_rate_hz.load(),
 		            obtained.freq);
+
 		mixer.sample_rate_hz = check_cast<uint16_t>(obtained.freq);
+		set_section_property_value("mixer",
```

Comment:
I would `assert(false)` on this block.  It should never be allowed to happen based on the flags we pass.  We're telling SDL the only thing we're willing to negotiate is the number of samples, not frequency.

```cpp
int sdl_allow_flags = 0;

if (negotiate) {
	// We allow chunk size changes (if supported), but don't allow
	// frequency changes because we've only seen them go beyond
	// SDL's desired 48 Khz limit
#if defined(SDL_AUDIO_ALLOW_SAMPLES_CHANGE)
	sdl_allow_flags |= SDL_AUDIO_ALLOW_SAMPLES_CHANGE;
#endif
}
```
I would actually re-consider these flags though (but not for backport).  If we tell SDL not to negotiate, it will re-sample our audio behind the screns from what we requested to whatever the underlying hardware and OS will accommodate.  SDL's re-sampler is kind of crap.  My one contribution to SDL was better integration of a third-party resampling library but that's off by default.  I believe Staging uses Speex for resampling which is much better.

Reply:
This is also contrary to what's documented in the config file help.  It says it will negotiate sample rate but it doesn't.

```cpp
bool_prop = sec_prop.Add_bool("negotiate", only_at_start, default_allow_negotiate);
bool_prop->Set_help(
        "Let the system audio driver negotiate possibly better sample rate and blocksize\n"
        "settings (%s by default).");
```

Reply:
On second thought, you don't have to add the assert.  It's a path we probably want to support in the future.  I believe it should never with the way the SDL flags are now though.

Reply:
> We're telling SDL the only thing we're willing to negotiate is the number of samples, not frequency.

@weirddan455 Huh, I kinda glossed over that detail. Good to know.

> I would actually re-consider these flags though (but not for backport). If we tell SDL not to negotiate, it will re-sample our audio behind the screns from what we requested to whatever the underlying hardware and OS will accommodate.

That's a bit crap. I'd expect to to simply fail with negotiation off if the driver doesn't support that requested rate, and with negotiation on, it should just find the closest driver-supported rate. 

> SDL's re-sampler is kind of crap. My one contribution to SDL was better integration of a third-party resampling library but that's off by default. My motivation for doing so was the built-in one was seg-faulting in some cases 🙄 I believe Staging uses Speex for resampling which is much better.

I'd really like disallow _any_ SDL-side resampling whatsoever. Is that possible with some flags? I want it to work as per above, and I want to keep all resampling business in our hands.

Yeah, I introduced Speex, it's very good quality, fast enough even on the Raspbery Pi 4, and has low latency, which is very important in emulators.

I found the Speex quality setting of 5 to be _almost_ transparent. I could only hear tiny differences with very intense A/B comparisons on very special material in studio headphones, and even then I wasn't 100% sure it wasn't just placebo or expectation bias...